File: serve.py
"""Static server for local development.

    python serve.py            # port 5173
    python serve.py 5178       # custom port

Binds to every interface so you can open the site on a phone on the same
Wi-Fi. The LAN address is printed on start.
"""
import http.server
import socket
import os
import sys

ROOT = os.path.dirname(os.path.abspath(__file__))
PORT = int(sys.argv[1]) if len(sys.argv) > 1 else 5173


class Handler(http.server.SimpleHTTPRequestHandler):
    def __init__(self, *args, **kwargs):
        super().__init__(*args, directory=ROOT, **kwargs)

    extensions_map = {
        **http.server.SimpleHTTPRequestHandler.extensions_map,
        ".js": "text/javascript",
        ".mjs": "text/javascript",
        ".css": "text/css",
        ".woff2": "font/woff2",
        ".svg": "image/svg+xml",
    }

    def end_headers(self):
        self.send_header("Cache-Control", "no-store")
        super().end_headers()

    def log_message(self, *args):
        pass


def lan_ip():
    """Best guess at this machine's address on the local network."""
    s = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
    try:
        s.connect(("10.255.255.255", 1))  # never actually sends anything
        return s.getsockname()[0]
    except Exception:
        return None
    finally:
        s.close()


class Server(http.server.ThreadingHTTPServer):
    # Threading matters: a single-threaded server blocks every other client
    # while one browser holds a keep-alive connection, which makes testing
    # on a phone while the desktop tab is open look like a dead server.
    allow_reuse_address = True
    daemon_threads = True


with Server(("0.0.0.0", PORT), Handler) as httpd:
    ip = lan_ip()
    print("\n  Ashmont Gin")
    print(f"  this machine   http://localhost:{PORT}")
    if ip:
        print(f"  same Wi-Fi     http://{ip}:{PORT}")
    print("\n  Ctrl+C to stop.\n")
    httpd.serve_forever()

File: drafts/04-heritage/serve.js
/**
 * Static server for local development. Node twin of serve.py.
 *
 *     npm start            # port 5173
 *     node serve.js 5178   # custom port
 *
 * Binds to every interface so you can open the site on a phone on the same
 * Wi-Fi. The LAN address is printed on start.
 */
import http from "node:http";
import fs from "node:fs";
import path from "node:path";
import os from "node:os";
import { fileURLToPath } from "node:url";

const ROOT = path.dirname(fileURLToPath(import.meta.url));
const PORT = Number(process.argv[2]) || 5173;

const TYPES = {
  ".html": "text/html; charset=utf-8",
  ".js": "text/javascript",
  ".mjs": "text/javascript",
  ".css": "text/css",
  ".json": "application/json",
  ".svg": "image/svg+xml",
  ".png": "image/png",
  ".jpg": "image/jpeg",
  ".jpeg": "image/jpeg",
  ".webp": "image/webp",
  ".avif": "image/avif",
  ".ico": "image/x-icon",
  ".woff2": "font/woff2",
  ".woff": "font/woff",
  ".mp4": "video/mp4",
  ".webm": "video/webm",
  ".glb": "model/gltf-binary",
};

/** Best guess at this machine's address on the local network. */
function lanIp() {
  for (const list of Object.values(os.networkInterfaces())) {
    for (const net of list ?? []) {
      if (net.family === "IPv4" && !net.internal) return net.address;
    }
  }
  return null;
}

function send(res, status, body, headers = {}) {
  res.writeHead(status, { "Cache-Control": "no-store", ...headers });
  res.end(body);
}

const server = http.createServer((req, res) => {
  const url = new URL(req.url, "http://localhost");
  let pathname = decodeURIComponent(url.pathname);
  if (pathname.endsWith("/")) pathname += "index.html";

  // Resolve inside ROOT so a "../" in the request cannot escape the folder.
  const file = path.join(ROOT, path.normalize(pathname));
  if (file !== ROOT && !file.startsWith(ROOT + path.sep)) {
    return send(res, 403, "Forbidden");
  }

  fs.stat(file, (err, stat) => {
    if (err || !stat.isFile()) return send(res, 404, "Not found");
    const type = TYPES[path.extname(file).toLowerCase()] ?? "application/octet-stream";
    res.writeHead(200, {
      "Content-Type": type,
      "Content-Length": stat.size,
      "Cache-Control": "no-store",
    });
    fs.createReadStream(file).pipe(res);
  });
});

server.listen(PORT, "0.0.0.0", () => {
  const ip = lanIp();
  console.log("\n  Ashmont Gin");
  console.log(`  this machine   http://localhost:${PORT}`);
  if (ip) console.log(`  same Wi-Fi     http://${ip}:${PORT}`);
  console.log("\n  Ctrl+C to stop.\n");
});
